stl/inc/ranges

```diff
@@ -620,6 +620,156 @@ namespace ranges {
         /* [[no_unique_address]] */ _Ty _Value{};
     };
 
+    // CLASS TEMPLATE ranges::_Non_propagating_cache
+    template <class _Ty>
+    concept _Objectable = is_object_v<_Ty>;
+
+    // clang-format off
+    template <_Objectable _Ty>
+    class _Non_propagating_cache { // a simplified optional that resets on copy / move
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept : _Dummy{} {}
+
+        _CONSTEXPR20_DYNALLOC ~_Non_propagating_cache() {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy = {};
+            }
+        }
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept {}
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept {
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Dummy   = {};
+                _Other._Engaged = false;
+            }
+        }
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache& operator=(const _Non_propagating_cache& _Other) noexcept {
+            if (_STD addressof(_Other) == this) {
+                return *this;
+            }
+
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy   = {};
+                _Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache& operator=(_Non_propagating_cache&& _Other) noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Dummy   = {};
+                _Engaged = false;
+            }
+
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Dummy   = {};
+                _Other._Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+        template <class _It>
+        _CONSTEXPR20_DYNALLOC _Ty& _Emplace_deref(_It& _Iter) noexcept(noexcept(_Ty(*_Iter))) /* strengthened */ {
+            if constexpr (!is_nothrow_constructible_v<_Ty, decltype(*_Iter)>) {

```

Comment:
Why is there a check here? Shouldn't we destruct whatever was engaged? In the vector case, this just leaves the allocations dangling. For the record, I believe this is causing the code to not compile after the latest fixes internally.

Reply:
Technically construct at should destroy the existing Element at that position

Reply:
I don't see that here: https://eel.is/c++draft/specialized.construct. It seems equivalent to placement new and that needs explicit destruction.

Reply:
You are absolutely correct. Seems I misremembered that part